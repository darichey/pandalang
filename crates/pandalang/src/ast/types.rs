// TODO: syntax for polymorphic types
#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Simple(String),
    Fun(Fun),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub from: Box<Type>,
    pub to: Box<Type>,
}
