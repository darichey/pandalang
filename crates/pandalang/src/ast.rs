#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub patt: Pattern,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub patt: Pattern,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
        }
        .to_string()
    }
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Pattern {
    Id {
        name: String,
        // typ: Type
    },
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Type {
    Base { name: String },
    Fun(Box<Type>, Box<Type>),
}
