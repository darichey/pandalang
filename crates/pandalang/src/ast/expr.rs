#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Expr {
    Int(Int),
    Str(Str),
    Unit,
    Bool(Bool),
    Var(Var),
    BinOp(BinOp),
    Let(Let),
    Fun(Fun),
    App(App),
    If(If),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Int {
    pub n: i64,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Str {
    pub s: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Bool {
    pub b: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Var {
    pub name: String,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct BinOp {
    pub left: Box<Expr>,
    pub right: Box<Expr>,
    pub kind: BinOpKind,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Fun {
    pub arg: String,
    pub body: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct App {
    pub fun: Box<Expr>,
    pub arg: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct If {
    pub check: Box<Expr>,
    pub then: Box<Expr>,
    pub els: Box<Expr>,
}

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
    Eql,
}

impl ToString for BinOpKind {
    fn to_string(&self) -> String {
        match self {
            BinOpKind::Add => "+",
            BinOpKind::Sub => "-",
            BinOpKind::Mul => "*",
            BinOpKind::Div => "/",
            BinOpKind::Eql => "==",
        }
        .to_string()
    }
}
