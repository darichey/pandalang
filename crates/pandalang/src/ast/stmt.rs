use super::{expr::Expr, types::Type};

#[derive(PartialEq, Eq, Debug, Clone)]
pub enum Stmt {
    Let(Let),
    Declare(Declare),
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Let {
    pub name: String,
    pub value: Box<Expr>,
    pub rec: bool,
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Declare {
    pub name: String,
    pub typ: Type,
}
