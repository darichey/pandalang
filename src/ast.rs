#[derive(PartialEq, Eq, Debug, Clone)]
pub enum ExprF<A> {
    Int {
        n: i64,
    },
    Var {
        name: String,
    },
    BinOp {
        left: A,
        right: A,
        kind: BinOpKind,
    },
    Fun{
        arg: String,
        body: A,
    },
    App{
        fun: A,
        arg: A,
    },
}

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Expr(pub Box<ExprF<Expr>>);

#[derive(PartialEq, Eq, Debug, Clone, Copy)]
pub enum BinOpKind {
    Add,
    Sub,
    Mul,
    Div,
}
