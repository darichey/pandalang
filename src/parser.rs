use lalrpop_util::{lexer::Token, ParseError};

use crate::ast::Expr;

lalrpop_mod!(pub grammar);

pub fn parse<'input>(
    s: &'input str,
) -> Result<Expr, ParseError<usize, Token<'input>, &'static str>> {
    grammar::ExprParser::new().parse(s)
}

#[cfg(test)]
mod tests {
    use crate::{ast::Expr, parser};

    fn parse(s: String) -> Expr {
        parser::parse(s.as_str()).unwrap()
    }

    #[test]
    fn parses() {
        insta::glob!("snapshot_inputs/**/*.panda", |path| {
            let source = parse(std::fs::read_to_string(&path).unwrap());
            insta::assert_debug_snapshot!(source);
        });
    }
}
