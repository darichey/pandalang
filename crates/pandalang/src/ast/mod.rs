use self::stmt::Stmt;

pub mod expr;
pub mod stmt;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
