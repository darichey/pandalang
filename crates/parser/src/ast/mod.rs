use self::stmt::Stmt;

pub mod expr;
pub mod stmt;
pub mod types;

#[derive(PartialEq, Eq, Debug, Clone)]
pub struct Program {
    pub stmts: Vec<Stmt>,
}
