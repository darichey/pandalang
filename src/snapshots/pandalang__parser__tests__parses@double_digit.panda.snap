---
source: src/parser.rs
expression: source
input_file: src/snapshot_inputs/int/double_digit.panda
---
Expr(
    Int {
        n: 10,
    },
)
